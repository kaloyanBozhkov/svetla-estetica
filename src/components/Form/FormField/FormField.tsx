// import commoon types
import type { ReactElement } from 'react'
import type { FormFieldProps } from 'common/types'

// import atoms
import Input from 'components/UI/Input/Input'

// import styles
import styles from './styles.module.scss'

export default function FormField({
    type,
    label,
    name,
    value,
    modifier,
    required,
    hasErrored,
    description,
    errorMsg,
    onBlur,
    onFocus,
    onChange,
}: FormFieldProps): ReactElement {
    switch (type) {
        // fall-through since text and apssword both use same input setup
        case 'textarea':
        case 'input':
            return (
                <div
                    className={styles.formField}
                    form-field={name}
                    form-type={type}
                    data-has-errored={hasErrored}
                >
                    <Input
                        type={type}
                        name={name}
                        label={label}
                        value={value}
                        modifier={modifier}
                        required={required}
                        onBlur={onBlur}
                        onFocus={onFocus}
                        onChange={onChange}
                    />
                    {errorMsg && <p className={styles.errorMsg}>{errorMsg}</p>}
                    {description && <p className={styles.desc}>{description}</p>}
                </div>
            )
        default:
            return <p>type {type} not configured</p>
    }
}
