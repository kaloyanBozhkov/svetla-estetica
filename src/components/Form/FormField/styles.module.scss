@import "scss/variables.scss";
@import "scss/mixins.scss";

.formField {
    > .desc {
        margin-top: 1.6rem;
        color: $black;
        font-size: 1.4rem;
        font-weight: 300;
    }

    > .errorMsg {
        margin-top: 0.5rem;
        color: $pink;
        font-size: 1.4rem;
        font-weight: 300;
    }

    label {
        animation-timing-function: ease-in-out;
        transform-origin: 0 0;
        animation-direction: alternate;
        color: $pink;
    }

    input {
        box-sizing: border-box;
        border-color: $pink;
    }

    &[data-has-errored="true"] {
        [data-is-active="false"] {
            label {
                @keyframes shake1 {
                    10%,
                    90% {
                        transform: translateY(-50%) translate3d(-1px, 0, 0);
                    }

                    20%,
                    80% {
                        transform: translateY(-50%) translate3d(2px, 0, 0);
                    }

                    30%,
                    50%,
                    70% {
                        transform: translateY(-50%) translate3d(-4px, 0, 0);
                    }

                    40%,
                    60% {
                        transform: translateY(-50%) translate3d(4px, 0, 0);
                    }
                }
                @include animate(shake1, 1s);
            }
        }

        [data-is-active="true"] {
            label {
                color: $pink;
            }
            input {
                @keyframes shake2 {
                    10%,
                    90% {
                        padding-left: 1rem;
                    }

                    20%,
                    80% {
                        padding-left: 1.4rem;
                    }

                    30%,
                    50%,
                    70% {
                        padding-left: 1rem;
                    }

                    40%,
                    60% {
                        padding-left: 1.8rem;
                    }
                }
                @include animate(shake2, 1s);

                color: $pink;
            }
        }
    }
}
