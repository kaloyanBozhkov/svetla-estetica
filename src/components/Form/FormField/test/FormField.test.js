import React from 'react'
import { shallow } from 'enzyme'

import FormField from '../FormField'

describe('testing FormField', () => {
    const formFieldConfigs = {
        input: {
            type: 'textarea',
            label: 'title',
            name: 'fieldName',
            value: '',
            modifier: undefined,
            required: true,
            hasErrored: false,
            errorMsg: undefined,
            description: undefined,
            onChange: jest.fn(),
            onFocus: jest.fn(),
            onBlur: jest.fn(),
        },
    }

    beforeEach(() => {
        formFieldConfigs.input.onFocus.mockReset()
        formFieldConfigs.input.onBlur.mockReset()
        formFieldConfigs.input.onChange.mockReset()
    })

    it('should render component', () => {
        const wrapper = shallow(<FormField {...formFieldConfigs.input} />)

        expect(wrapper).toMatchSnapshot()
    })

    it('should render text input with error msg if property set', () => {
        const wrapper = shallow(<FormField {...formFieldConfigs.input} errorMsg="some error" />)

        expect(wrapper.find('.errorMsg').text()).toEqual('some error')
    })

    it('should render text input with description if property set', () => {
        const wrapper = shallow(
            <FormField {...formFieldConfigs.input} description="some description" />
        )

        expect(wrapper.find('.desc').text()).toEqual('some description')
    })

    it('should render fallback text tag', () => {
        const wrapper = shallow(<FormField {...formFieldConfigs.input} type="nonexistant" />)

        expect(wrapper.find('p').text()).toEqual('type nonexistant not configured')
    })
})
