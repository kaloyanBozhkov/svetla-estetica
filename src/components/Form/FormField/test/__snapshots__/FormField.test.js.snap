// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing FormField should render component 1`] = `
<div
  className="formField"
  data-has-errored={false}
  form-field="fieldName"
  form-type="textarea"
>
  <Input
    label="title"
    name="fieldName"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    required={true}
    type="textarea"
    value=""
  />
</div>
`;
